## SESSION I: Data Structures

***

There is a dataset "cities.xls" and the homework asks to get insights by making queries. Import the dataset and have a glimpse at the data:

In [225]:
import pandas as pd
import os

folderName = '/Users/liyuanzhang/Google Drive/17Q0_PubPol599C'
fileName = '1.cities.xlsx'
fileExcel = os.path.join(folderName,fileName)

cities = pd.read_excel(fileExcel)
cities.head()

city  region  visits2013  visits2015
0       Bangkok    Asia       15.98       18.24
1         Dubai    Asia        9.89       14.26
2     Hong Kong    Asia        8.72        8.66
3      Istanbul  Europe       10.37       12.56
4  Kuala Lumpur    Asia        9.20       11.12

<br>
**I-1. What was the most visited city in 2015?**

In [4]:
cities[cities.visits2015 == max(cities.visits2015)].city

5    London
Name: city, dtype: object

**I-2. What was the least visited city in Asia in 2013?**

In [7]:
asia = cities[cities.region == 'Asia']
asia[asia.visits2015 == min(asia.visits2015)].city

2    Hong Kong
Name: city, dtype: object

**I-3. What are the three least visited cities in Asia in 2015?**

In [8]:
asiaAsc = asia.sort_values(by = ['visits2015'], ascending = True)
asiaAsc.iloc[[0, 1, 2]]

city region  visits2013  visits2015
2     Hong Kong   Asia        8.72        8.66
8         Seoul   Asia         NaN       10.35
4  Kuala Lumpur   Asia        9.20       11.12

**I-4. What are the most and least visited city in Asia in 2013?**

In [9]:
# Most Visited
asia[asia.visits2013 == max(asia.visits2013)].city

0    Bangkok
Name: city, dtype: object

In [10]:
# Least Visited
asia[asia.visits2013 == min(asia.visits2013)].city

2    Hong Kong
Name: city, dtype: object

<br><br>
## SESSION II: Control of Execution

***

**II-1 Play with for loop. Check out numpy data type**

The orginial for loop:

In [13]:
import numpy as np
from math import sqrt

values=[9,None,np.nan, '1000',-100, 144,-72]
for value in values: # notice the order of 'IFs'
    if value==None: # condition1
        print ('missing values as input')
        continue
    if isinstance(value, str): #condition2
        print ('string as input')
        continue
    if value < 0: # condition3
        print ('negative value as input')
        continue
    print (sqrt(value), 'is the root of ',value)  

3.0 is the root of  9
missing values as input
nan is the root of  nan
string as input
negative value as input
12.0 is the root of  144
negative value as input


*II-1-1 Change the line 'value==None' for **np.isnan**(value) and explain what happens. You want to change the values in the list to see the different outputs.*

np.isnan cannot work with None.

In [87]:
import numpy as np
values = [9, np.nan, '1000', -100, 144, -72]
for value in values:
    try:
        if np.isnan(value):
            print('missing values as input')
            continue
    except TypeError:
        continue
    if isinstance(value, str):
        print('string as input')
        continue
    if value < 0:
        print('negative value as input')
        continue
    print(sqrt(value), 'is the root of', value)

3.0 is the root of 9
missing values as input
negative value as input
12.0 is the root of 144
negative value as input


*II-1-2 Change the order of all the conditions and explain what happens.*

If the "value < 0" condition is before the other two, the non-numeric data type (None, np.nan, '1000') will break the for loop before stepping into their evaluation.

In [92]:
import numpy as np

values = [9, None, np.nan, '1000', -100, 144, -72]
for value in values:
    try:
        if value < 0:
            print('negative value as input')
            continue
    except TypeError:
        continue
    if isinstance(value, str):
        print('string as input')
        continue
    if pa.isnull(value):
        print('missing values as input')
        continue
    print(sqrt(value), 'is the root of', value)

3.0 is the root of 9
missing values as input
negative value as input
12.0 is the root of 144
negative value as input


*II-1-3 The function **isnull** and **notnull** from pandas are very helpful to determine if you have a missing value, change the first condition to use **isnull**.*

In [89]:
import numpy as np
import pandas as pa
values = [9, None, np.nan, '1000', -100, 144, -72]
for value in values:
    if pa.isnull(value):
        print('missing values as input')
        continue
    if isinstance(value, str):
        print('string as input')
        continue
    if value < 0:
        print('negative value as input')
        continue
    print(sqrt(value), 'is the root of', value)

3.0 is the root of 9
missing values as input
missing values as input
string as input
negative value as input
12.0 is the root of 144
negative value as input


**II-2. Queries**

*II-2-1. Change the code below to find the youngest age, and inform who has that age.*

The code:

In [98]:
folderName = '/Users/liyuanzhang/Google Drive/17Q0_PubPol599C'
fileName = '1.people.xlsx'
fileExcel = os.path.join(folderName,fileName)

people = pd.read_excel(fileExcel)

numRows,numCols=people.shape
oldestAge=0
positionsOfOldest=[]
countOldest=0

# get the oldest age
for position in range(numRows):
    age = people.loc[position,'ages']
    if oldestAge<age:
        oldestAge=age

# get position of the oldest age
for position in range(numRows):
    age = people.loc[position,'ages']
    if oldestAge==age:
        countOldest=countOldest+1
        positionsOfOldest.append(position)

Make the change to find the youngest age and the person:

In [116]:
youngestAge = 200
positionsOfYoungest=[]

for position in range(numRows):
    age = people.loc[position,'ages']
    if youngestAge > age:
        youngestAge = age

for position in range(numRows):
    age = people.loc[position,'ages']
    if youngestAge == age:
        positionsOfYoungest.append(position)

In [117]:
# The youngest age
youngestAge

25

In [119]:
# The youngest peopole
people.iloc[positionsOfYoungest]

cod  ages state  education
2    A3    25    DC          3
7    A8    25    NY          2
19  A20    25    FL          1

*II-2-2. Redo the first two questions in the homework of session 1 using if and for (and related functions).*

*REDO I-1. What was the most visited city in 2015?*

In [141]:
numRows, numCols = cities.shape
max15 = 0
max15p = []

for position in range(numRows):
        visits = cities.loc[position,'visits2015']
        if visits == None:
            continue
        elif max15 < visits:
            max15p = [position]
            max15 = visits
        elif max15 == visits:
            max15p.append(position)

In [140]:
# The most visited city
cities.iloc[max15p]

city  region  visits2013  visits2015
5  London  Europe       15.96       18.82

*REDO I-2. What was the least visited city in Asia in 2013?*

In [142]:
min13 = 1e12
min13p = []

for position in range(numRows):
        visits = cities.loc[position, 'visits2013']
        region = cities.loc[position, 'region']
        
        if region != "Asia" or visits == None:
            continue
        elif min13 > visits:
            min13p = [position]
            min13 = visits
        elif min13 == visits:
            min13.append(position)

In [143]:
# The least visited city
cities.iloc[min13p]

city region  visits2013  visits2015
2  Hong Kong   Asia        8.72        8.66

*II-2-3. Redo the third question in the homework of session 1 using if and for (and related functions).*

*REDO I-3. What are the three least visited cities in Asia in 2015?*

In [272]:
cities1 = cities.copy()

min15n = []
min15cities = []

for n in range(3):
    min15 = 1e12
    min15p = []
    
    for position in range(len(cities1)):
        visits = cities1.loc[position].visits2015
        region = cities1.loc[position].region

        if region != "Asia" or visits == None:
            continue
        elif min15 > visits:
            min15p = [position]
            min15 = visits
        elif min15 == visits:
            min15p.append(position)
            
    min15n.append(min15)
    min15cities.append(cities1.iloc[min15p].city.tolist())
    cities1 = cities1.drop(min15p).reset_index(drop = True)

In [275]:
pd.DataFrame({'city': min15cities, 'visits2015': min15n})

city  visits2015
0     [Hong Kong]        8.66
1         [Seoul]       10.35
2  [Kuala Lumpur]       11.12

**II-3. Comprehension**

In [14]:
values=[9,25,49,-121]
rootsInDict={value:(sqrt(value) if value > 0 else None) 
             for value in values}
rootsInDict

{-121: None, 9: 3.0, 25: 5.0, 49: 7.0}

*II-3-1 Create a list of tuples, with the values from rootsInDict.*

In [15]:
[(key, value) for key, value in rootsInDict.items()]

[(9, 3.0), (-121, None), (49, 7.0), (25, 5.0)]

*II-3-2 Create a list with the first 10 even numbers, then using comprehension create a list with each of those values to the power of two, and using comprehension another list with each of those same values with their square root; then another comprehension to create a list of tuples combining the three lists, with each tuple holding the triplet: (number, numer to power of two, squared root of number.*

In [17]:
lseven = [2 * (i + 1) for i in range(10)]
lseven2 = [value ** 2 for value in lseven]
lsevensqrt = [sqrt(value) for value in lseven] 
lstp = list(zip(lseven, lseven2, lsevensqrt)) 
lstp

[(2, 4, 1.4142135623730951),
 (4, 16, 2.0),
 (6, 36, 2.449489742783178),
 (8, 64, 2.8284271247461903),
 (10, 100, 3.1622776601683795),
 (12, 144, 3.4641016151377544),
 (14, 196, 3.7416573867739413),
 (16, 256, 4.0),
 (18, 324, 4.242640687119285),
 (20, 400, 4.47213595499958)]

*II-3-3 Create a dictionary of dictionaries with the three lists created initially in the previous exercise. The dict of dicts should be named 'numPowRoot', and if you write: numPowRoot[10], your output should be {'power2': 100, 'root2': 3.1622776601683795}*

In [20]:
numPowRoot = {item[0] : ({'power2': item[1], 'root2': item[2]}) 
for item in list(zip(lseven, lseven2, lsevensqrt))}

numPowRoot[10]

{'power2': 100, 'root2': 3.1622776601683795}

<br><br>
## SESSION III: Building Functions

***

**III-1. Change the function below to create the function `riseToPowerPlus`, which gives a good answer even when the power is negative.**

The original function:

In [24]:
def riseToPower(base,exponent=2):
    result=1
    
    if exponent > 0:
        for time in range(1,exponent+1):
            result=result*base
    
    return(result)

In [39]:
# Testing
[riseToPower(9), riseToPower(9, 3), riseToPower(9, 0)]

[81, 729, 1]

The improved one:

In [27]:
def riseToPowerPlus(base, exponent = 2):
    result = 1
    if exponent > 0:
        for time in range(1, exponent+1):
            result = result * base
    elif exponent < 0:
        for time in range(1, abs(exponent)+1):
            result = result * base
        result = 1/result
    return(result)

In [40]:
# Testing
riseToPowerPlus(2, -2)

0.25

**III-2. Play with the `factor` function**

First have a look at the function:

In [31]:
def factors(number):
    factorsList=[]
    
    for i in range(1, number + 1):
        if number % i == 0:
            factorsList.append(i)

    return factorsList

In [41]:
# Testing
factors(36)

[1, 2, 3, 4, 6, 9, 12, 18, 36]

*III-2-1. Change the function ‘factors’ to reduce the amount of iterations in the for loop and still get the factors shown above.*

In [34]:
def factors(number):
    factorsList = []
    
    from math import sqrt
    for i in range(1, int(sqrt(abs(number))) + 1):
        if number % i == 0:
            factorsList.append(i)
            
    import numpy as np
    factorsList += list(number / np.array(factorsList))
    factorsList = list(set(factorsList))
    
    factorsList = [int(e) for e in factorsList]
    return sorted(factorsList)

In [42]:
# Testing
factors(36)

[1, 2, 3, 4, 6, 9, 12, 18, 36]

*III-2-2. Change your new function ‘factors’ to avoid negative values or zeros as input; your code should return a message if an invalid valid was input.*

In [38]:
def factors(number):
    
    if number <= 0:
        print('Zero or negative value detected. Program stopped.')
        exit
    else:
        factorsList = []
        
        from math import sqrt
        for i in range(1, int(sqrt(abs(number))) + 1):
            if number % i == 0:
                factorsList.append(i)
                
        import numpy as np
        factorsList += list(number / np.array(factorsList))
        factorsList = list(set(factorsList))
        
        factorsList = [int(e) for e in factorsList]
        return sorted(factorsList)

In [43]:
# Testing
[factors(0), factors(-20)]

Zero or negative value detected. Program stopped.
Zero or negative value detected. Program stopped.


[None, None]

*III-2-3. Change your newest function ‘factors’ to accept only positive integer values; your code should return a message if an invalid valid was input.*

In [44]:
def factors(number):
    
    if number <= 0 or number % 1 !=0:
        print('Not a positive integer. Program stopped.')
        exit
    else:
        factorsList = []
        
        from math import sqrt
        for i in range(1, int(sqrt(abs(number))) + 1):
            if number % i == 0:
                factorsList.append(i)
                
        import numpy as np
        factorsList += list(number / np.array(factorsList))
        factorsList = list(set(factorsList))
        
        factorsList = [int(e) for e in factorsList]
        return sorted(factorsList)

In [46]:
# Testing
test = [1, 8, -1, 0.3]
list(map(factors, test))

Not a positive integer. Program stopped.
Not a positive integer. Program stopped.


[[1], [1, 2, 4, 8], None, None]

**III-3. Make a function that reads two vectors and returns a data frame with those vectors and extra columns with their sum, difference, multiplication and division.**

In [47]:
def calDF(aList, bList):
    import pandas as pd
    
    l = len(aList)
    
    sumList = [aList[i] + bList[i] for i in range(l)]
    difList = [aList[i] - bList[i] for i in range(l)]
    multList = [aList[i] * bList[i] for i in range(l)]
    divList = [aList[i] / bList[i] for i in range(l)]
        
    answerAsDicts = {'NumberA': aList, 'NumberB': bList, 
    'Sum': sumList, 'Difference': difList,
    'Multiplication': multList, 'Division': divList}
    
    return pd.DataFrame(answerAsDicts)

In [48]:
# Testing
aLs = [1, 2, 3]
bLs = [3, 4, 5]

calDF(aLs, bLs)

Difference  Division  Multiplication  NumberA  NumberB  Sum
0          -2  0.333333               3        1        3    4
1          -2  0.500000               8        2        4    6
2          -2  0.600000              15        3        5    8

**III-4. How should you organize the input so that you get `double()` to work as it did with R vectors?**

The double function:

In [49]:
def double(x):
    return 2*x

In [51]:
myList=[1,2,3]
[double(e) for e in myList]

[2, 4, 6]

**III-5. Applying function `strangerF` to composite structures.**

*III-5-1. Understand what this function is doing*

Function `strangerF` will take a series of numbers (vector/data.frame) and return one number as the result.

In [53]:
def strangerF(x):
    answer=0
    change=True
    for value in x:
        if (change):
            answer+=value
        else:
            answer-=value
    
        change= not change 

    return answer

*III-5-2. Use the function above with the the file ‘dataforstrangeF.csv’, find out which way of “applying the function”" works, that is, report whether you can apply the function directly to the data frame, to a subset of a data frame, to a column, or to a cell, and whether apply, mapply, Map, Filter, lapply work with the function or not in the data set given.*

Import and understand the data:

In [55]:
folderName = '/Users/liyuanzhang/Google Drive/17Q0_PubPol599C'
fileName = '1.dataforstrangeF.csv'
fpth = os.path.join(folderName,fileName)

dataCsv = pd.read_csv(fpth)

In [59]:
dataCsv.head()

vals1  vals2  vals3
0     A      9    100
1     B      1      1
2     C      1      1
3     A      1      1
4     B      1      1

Subset `dataCsv` to different data types and apply the functions to each of them:

In [62]:
# Data frame containing only numeric values
numDF = dataCsv.iloc[:, 1:3]

# Number series (1 column)
numSeries = dataCsv.iloc[:, 1]

# One column as data frame
vals2DF = dataCsv.iloc[:, [1]]

dataCsv.iloc[:, 1]

Now apply the `strangerF` function and check out how it works with different data types:

1) TypeError: unsupported operand type(s) for +=: 'int' and 'str'

```
strangerF(dataCsv)
list(map(strangerF, numDF)) 
```

  *Note: `map` will apply the function to the column names of num DF*

2) TypeError: 'numpy.int64' object is not iterable

```
list(map(strangerF, numSeries))
numSeries.apply(strangerF)
numDF.applymap(strangerF)
```
  *Note: `map` cannot apply a function with iteration to not iterable object, but the data frame ".apply" method will work.*

Below are conditions when the function work, along with the output:

In [67]:
numDF.apply(strangerF, axis = 0)

vals2      9
vals3    100
dtype: int64

In [68]:
numDF.apply(strangerF, axis = 1)

0   -91
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
dtype: int64

In [69]:
vals2DF.apply(strangerF)

vals2    9
dtype: int64

In [73]:
vals2DF.apply(strangerF, axis = 1)

0    9
1    1
2    1
3    1
4    1
5    1
6    1
7    1
8    1
dtype: int64